### PR TITLE
Generic monster attacks, take 2: "Library of Attacks"

### PR DESCRIPTION
**Pull request by Coolthulhu**
This got pretty giant. Recreating PR instead of reopening #17679 because it got some giant updates.

Contains the generic melee attacks as in #17679 but also a "library of monster attacks" which works as follows:
* Monster attacks specified in base context (ie. not in a monster) with type "monster_attack" are written down into `MonsterGenerator::attack_map`
* In monster context, a monster can either request one of the base types ("leap", "bite" etc.) or "monster_attack", which covers library ones
* A monster can request a library type and make inline modifications to it, in case something tiny needs to be changes - for example cooldown
* Due to the limitations of current implementation, nested library types are NOT supported at the moment
* Hardcoded mattacks are wrapped into a structure similar to that of `iuse` functions
* Monster attack "type" renamed to "attack_type", as "type" is a reserved keyword for init functions
* Old definitions are supported for now, but will cause warnings in `test_mode` (`--check-mods`)

For example:
```
{
    "type": "monster_attack",
    "attack_type": "bite",
    "id": "nibble",
    "move_cost": 10
}
...
{
    "type": "monster",
    "id": "nibblerung"
    ...
    "special_attacks": [ { "id": "nibble" } ]
}
```

It is also possible to use the array definition (`"special_attacks": [ [ "nibble", 2 ] ]`).

EDIT: Added a doc: https://github.com/CleverRaven/Cataclysm-DDA/wiki/Monster-special-attacks

### CODE REVIEW

File: src/monstergenerator.cpp [hunk 1]
```diff
@@ -665,46 +672,121 @@ m_flag MonsterGenerator::m_flag_from_string( std::string flag ) const
...
     }
 
-    special_attacks_names.push_back( type );
+    new_attack->load( obj, src );
+    return mtype_special_attack( new_attack );
+}
+
+void mattack_actor::load( JsonObject &jo, const std::string &src )
+{
+    bool strict = src == "core";
```
Comment (codemime) [inline src/monstergenerator.cpp]: This check doesn't work anymore. The core mod is `"dda"`.

File: src/mattack_actors.cpp [hunk 2]
```diff
@@ -131,15 +132,25 @@ bool leap_actor::call( monster &z ) const
...
 {
     damage_max_instance = damage_instance::physical( 9, 0, 0, 0 );
     min_mul = 0.5f;
     max_mul = 1.0f;
     move_cost = 100;
 }
 
-void bite_actor::load( JsonObject &obj )
+void load_if_available( std::string &to, JsonObject &obj, const std::string &id,
+                        const std::string &def )
```
Comment (codemime) [inline src/mattack_actors.cpp]: This looks like a reinvented `optional( ..., translated_string_reader )`.

**Comment by DangerNoodle** [comment]
I assume this will warrant reporting on https://github.com/CleverRaven/Cataclysm-DDA/issues/19376 when it is merged.

File: src/mattack_actors.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/mattack_actors.cpp]: How to get it without generic factory?

Comment (codemime) [inline src/mattack_actors.cpp]: `optional()` doesn't depend on it.

**Comment by chaosvolt** [comment]
Eh? Sure, I guess. Wait, JSON API? :V

**Comment by DangerNoodle** [comment]
See issue mentioned by me above, reporting major JSON changes.

**Comment by Coolthulhu** [comment]
Should be working now.

**Pull request merged**